pkg/gardenlet/controller/federatedseed/network-policy/namespace_reconciler.go

```diff
@@ -0,0 +1,85 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package network_policy
+
+import (
+	"context"
+
+	"github.com/gardener/gardener/pkg/gardenlet/controller/federatedseed/network-policy/helper"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	kubecorev1listers "k8s.io/client-go/listers/core/v1"
+	kubenetworkingv1listers "k8s.io/client-go/listers/networking/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// namespaceReconciler implements the reconcile.Reconcile interface for namespace reconciliation.
+type namespaceReconciler struct {
+	ctx                 context.Context
+	log                 *logrus.Entry
+	seedClient          client.Client
+	seedName            string
+	endpointsLister     kubecorev1listers.EndpointsLister
+	networkPolicyLister kubenetworkingv1listers.NetworkPolicyLister
+}
+
+// newNamespaceReconciler returns the new namespace reconciler.
+func newNamespaceReconciler(ctx context.Context, seedLogger *logrus.Entry, seedClient client.Client, endpointsLister kubecorev1listers.EndpointsLister, networkPolicyLister kubenetworkingv1listers.NetworkPolicyLister, seedName string) reconcile.Reconciler {
+	return &namespaceReconciler{
+		ctx:                 ctx,
+		seedClient:          seedClient,
+		endpointsLister:     endpointsLister,
+		networkPolicyLister: networkPolicyLister,
+		log:                 seedLogger,
+		seedName:            seedName,
+	}
+}
+
+// Reconcile reconciles shoot namespaces in order to create the "allowed-to-seedName-apiserver" Network Policy
+func (r *namespaceReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	logrus.Infof("Updating NetworkPolicy %q for namespace %q on seed %q", helper.AllowToSeedAPIServer, request.Name, r.seedName)
+
+	kubernetesEndpoints, err := r.endpointsLister.Endpoints(corev1.NamespaceDefault).Get("kubernetes")
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	egressRules := helper.GetEgressRules(kubernetesEndpoints)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// avoid unnecessary & duplicate NetworkPolicy updates
+	policy, err := r.networkPolicyLister.NetworkPolicies(request.Name).Get(helper.AllowToSeedAPIServer)
+	if client.IgnoreNotFound(err) != nil {
+		return reconcile.Result{}, err
+	}
+	if policy != nil {
+		if apiequality.Semantic.DeepEqual(policy.Spec.Egress, egressRules) {
+			return reconcile.Result{}, nil
+		}
+	}
```

Comment:
Why is this needed? AFAIK this is handled by the `CreateOrUpdate` function that you call in your helper, or?

Reply:
There is no need to call that function in the first place if you don't plan on updating it.

Reply:
Why not? As said, it already performs the exact same check like this one here. If the mutation function does not mutate the object, nothing will happen. Hence, we could save these lines here.

Reply:
`CreateOrUpdate` would do a `GET` request in this case? Which is not needed.

Reply:
The [Reader](https://github.com/gardener/gardener/blob/d1ac77f7302ba5011842f45fba976d786316249b/vendor/sigs.k8s.io/controller-runtime/pkg/client/interfaces.go#L105) in the client passed to `CreateOrUpdate` is not backed by a `cache`

Reply:
Yes, correct. But given that @tim-ebert is right now in the finalization of a PR that introduces the usage of cached clients everywhere in Gardener (see https://github.com/tim-ebert/gardener/tree/enh/gardenlet-cached-client, afaik it should be ready next week or so) I would not be too much concerned about it.
Anyways, I'm also okay with keeping it for now, just wanted to bring it up. Probably @tim-ebert can remove it again as part of his PR.

Reply:
Additionally, even though it is only a minor thing: there would be duplicate log entries for "Successfully updated NetworkPolicy xy". At the moment these logs are only written when there were changes applied. I somehow like having a clean log vs. saving a few lines.